src/utils.rs

```diff
@@ -1,17 +1,170 @@
+
+// Only compressed pubkeys are supported for silent payments
+const COMPRESSED_PUBKEY_SIZE: usize = 33;
+
+pub struct VinData {
+    pub script_sig: Vec<u8>,
+    pub txinwitness: Vec<Vec<u8>>,
+    pub prevout: Vec<u8>,
+}
+
+// script templates for inputs allowed in BIP352 shared secret derivation
+pub fn is_p2tr(spk: &[u8]) -> bool {
+    matches!(spk, [OP_1, OP_PUSHBYTES_32, ..] if spk.len() == 34)
+}
+
+fn is_p2wpkh(spk: &[u8]) -> bool {
+    matches!(spk, [OP_0, OP_PUSHBYTES_20, ..] if spk.len() == 22)
+}
+
+fn is_p2sh(spk: &[u8]) -> bool {
+    matches!(spk, [OP_HASH160, OP_PUSHBYTES_20, .., OP_EQUAL] if spk.len() == 23)
+}
+
+fn is_p2pkh(spk: &[u8]) -> bool {
+    matches!(spk, [OP_DUP, OP_HASH160, OP_PUSHBYTES_20, .., OP_EQUALVERIFY, OP_CHECKSIG] if spk.len() == 25)
+}
+
+sha256t_hash_newtype! {
+    pub struct InputsTag = hash_str("BIP0352/Inputs");
+
+    /// BIP0352-tagged hash with tag \"Inputs\".
+    ///
+    /// This is used for computing the inputs hash.
+    #[hash_newtype(forward)]
+    pub struct InputsHash(_);
+
+    pub struct LabelTag = hash_str("BIP0352/Label");
+
+    /// BIP0352-tagged hash with tag \"Label\".
+    ///
+    /// This is used for computing the label tweak.
+    #[hash_newtype(forward)]
+    pub struct LabelHash(_);
+
+    pub struct SharedSecretTag = hash_str("BIP0352/SharedSecret");
+
+    /// BIP0352-tagged hash with tag \"SharedSecret\".
+    ///
+    /// This hash type is for computing the shared secret.
+    #[hash_newtype(forward)]
+    pub struct SharedSecretHash(_);
+}
+
+impl InputsHash {
+    pub fn from_outpoint_and_A_sum(
+        smallest_outpoint: &Vec<u8>,
+        A_sum: &PublicKey,
+    ) -> InputsHash {
+        let mut eng = InputsHash::engine();
+        eng.input(&smallest_outpoint);
+        eng.input(&A_sum.serialize());
+        InputsHash::from_engine(eng)
+    }
+    pub fn to_scalar(self) -> Scalar {
+        // This is statistically extremely unlikely to panic.
+        Scalar::from_be_bytes(self.to_byte_array()).expect("hash value greater than curve order")
+    }
+}
+
+impl LabelHash {
+    pub fn from_b_scan_and_m(
+        b_scan: SecretKey,
+        m: u32,
+    ) -> LabelHash {
+        let mut eng = LabelHash::engine();
+        eng.input(&b_scan.secret_bytes());
+        eng.input(&m.to_be_bytes());
+        LabelHash::from_engine(eng)
+    }
+    pub fn to_scalar(self) -> Scalar {
+        // This is statistically extremely unlikely to panic.
+        Scalar::from_be_bytes(self.to_byte_array()).expect("hash value greater than curve order")
+    }
+}
+
+impl SharedSecretHash {
+    pub fn from_ecdh_and_k(
+        ecdh: &PublicKey,
+        k: u32,
+    ) -> SharedSecretHash {
+        let mut eng = SharedSecretHash::engine();
+        eng.input(&ecdh.serialize());
+        eng.input(&k.to_be_bytes());
+        SharedSecretHash::from_engine(eng)
+    }
+}
+
+pub fn get_pubkey_from_input(vin: &VinData) -> Result<Option<PublicKey>, Error> {
```

Comment:
Actually I'm wondering, do we really need this helper here? Wouldn't it be easier to let this function out of the sp-lib itself and handle pubkey parsing one layer above in an hypothetical sp-wallet lib? We would be able to use rust-bitcoin helper function.

Reply:
@Sosthene00 I go back and forth on this one :sweat_smile:. I am leaning towards keeping it (or something like it) here because public key extraction is not strictly a wallet problem: scanners/indexers need to be able to extract pubkeys from a transaction, so a function which takes `CTxIn`s (transaction inputs) and extracts public keys makes sense to me.

For the sender, it seems less relevant, since it is a wallet problem i.e. a coin selection problem. Presumably, the sender already has access the public keys without needing to "extract" them.

Reply:
Moved to `tests` so this is resolved